### ISSUE TITLE
BatchNorm1d crashes with batch size of 1 in DataParallel mode

### ISSUE THREAD

**Issue by cfd1** [pyg-team/pytorch_geometric#829]
I have been trying to run a network with batch normalisation from torch.nn after the graph convolutions and global pooling. When running on multiple GPUs it crashes as some batches are of a single graph. I was using the ModelNet10 dataset with torch manual seed 0.

Proposed scatter algorithm (pull request #828) to ensure each device has 2 graphs present to avoid batch norm from crashing. It also improves the balance.

I have tested this with 4 GPUs with the following batch sizes 2 (removed batch norm for this test), 4, 6, 8 and 32.

An example of the improvement for a batch size of 32, showing the first 3 data lists.
Batch 0
Original device allocation: [0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 3, 3, 3]
Original graphs per device: [ 5, 11, 13, 3]
Original total nodes per device: [114166, 59242, 99157, 62128]
Proposed device allocation: [0, 1, 2, 3, 3, 2, 1, 3, 2, 1, 3, 0, 2, 1, 3, 2, 0, 1, 3, 1, 0, 2, 3, 0, 1, 2, 3, 0, 1, 3, 2, 1]
Proposed graphs per new_device: [6, 9, 8, 9]
Proposed total nodes per device: [83457, 83759, 83747, 83730]

Standard deviations
Original Nodes per device: 27268.0
Proposed Nodes per device: 144.7
Original Graphs per device: 4.8
Proposed Graphs per device: 1.4

Batch 1
Original device allocation: [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 2, 2, 3, 3, 3]
Original graphs per device: [16, 6, 7, 3]
Original total nodes per device: [143070, 110554, 147301, 172038]
Proposed device allocation: [0, 1, 2, 3, 3, 2, 1, 2, 1, 3, 1, 2, 3, 3, 1, 2, 3, 1, 2, 3, 1, 2, 3, 1, 2, 3, 1, 2, 0, 1, 3, 2]
Proposed graphs per new_device: [ 2, 10, 10, 10]
Proposed total nodes per device: [169202, 134669, 134480, 134612]

Standard deviations
Original Nodes per device: 25260.1
Proposed Nodes per device: 17307.7
Original Graphs per device: 5.6
Proposed Graphs per device: 4.0

Batch 2
Original device allocation: [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 2, 2, 2, 3, 3, 3, 3, 3, 3, 3])
Original graphs per device: [11, 6, 8, 7]
Original total nodes per device: [61676, 27303, 87470, 55921]
Proposed device allocation: [0, 1, 2, 3, 3, 2, 1, 2, 3, 1, 1, 3, 2, 2, 3, 1, 2, 1, 3, 2, 3, 1, 2, 3, 1, 2, 1, 3, 0, 3, 2, 1]
Proposed graphs per new_device: [ 2, 10, 10, 10]
Proposed total nodes per device: [59041, 57805, 57788, 57736]

Standard deviations
Original Nodes per device: 24688.6
Proposed Nodes per device: 633.0
Original Graphs per device: 2.2
Proposed Graphs per device: 4.0

Batch 3
Original device allocation: [0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3]
Original graphs per device: [ 3, 8, 11, 10]
Original total nodes per device: [41879, 34950, 24246, 32947]
Proposed device allocation: [0, 1, 2, 3, 3, 2, 3, 2, 3, 2, 3, 1, 2, 3, 2, 1, 3, 2, 1, 3, 2, 1, 3, 1, 2, 3, 1, 2, 0, 2, 1, 3]
Proposed graphs per new_device: [ 2, 8, 11, 11]
Proposed total nodes per device: [36336, 32729, 32580, 32377]

Standard deviations
Original Nodes per device: 7263.0
Proposed Nodes per device: 1892.5
Original Graphs per device: 3.6
Proposed Graphs per device: 4.2

As you can see the balance in the above cases.

The code could be improved, but it fixes the issue I am having. Any feedback would be welcome.

Best Regards

Adam

**Comment by cfd1** [comment]
Sorry jumped the gun, it needs more work.

It appears to run slower. 
**Original**
Epoch 1: 81s
Epoch 2: 71s
**Proposed**
Epoch 1: 111s
Epoch 2: 112s

And has memory issues now crashes at epoch 4. I'll try and find the mistake and update the request

**Comment by cfd1** [comment]
OK I think I've fixed the memory issue and that has sped it up. 
**Proposed**
Epoch 1: 95s
Epoch 2: 91s
I'll look into the speed issue.

Do I need to redefine the gather function since they are no longer in order?

**Comment by rusty1s** [comment]
This seems super interesting. Thank you for your contribution. I currently do not have time to look closer at this, but I will :)

**Comment by cfd1** [comment]
I have found the issues with the implementation and it is working now.
I have implemented a gather function to return the output in the same order as the input data.

**Comment by cfd1** [comment]
I profiled it last night and it shows that there is some room for improvement. However I did get a small 2% speed up in CUDA time for 5 epoch of training using a batch size of 32 on 4 GPUs.
